winsup/cygwin/include/cygwin/exit_process.h

```diff
@@ -0,0 +1,121 @@
+
+#include <tlhelp32.h>
+
+/**
+ * Terminates the process corresponding to the process ID and all of its
+ * directly and indirectly spawned subprocesses.
+ */
+static int
+terminate_process_tree(HANDLE main_process, int exit_code)
+{
+  HANDLE snapshot = CreateToolhelp32Snapshot(TH32CS_SNAPPROCESS, 0);
+  PROCESSENTRY32 entry;
+  DWORD pids[16384];
+  int max_len = sizeof(pids) / sizeof(*pids), i, len, ret = 0;
+  pid_t pid = GetProcessId(main_process);
+
+  pids[0] = (DWORD) pid;
+  len = 1;
+
+  /*
+   * Even if Process32First()/Process32Next() seem to traverse the
+   * processes in topological order (i.e. parent processes before
+   * child processes), there is nothing in the Win32 API documentation
+   * suggesting that this is guaranteed.
+   *
+   * Therefore, run through them at least twice and stop when no more
+   * process IDs were added to the list.
+   */
+  for (;;)
+    {
+      int orig_len = len;
+
+      memset(&entry, 0, sizeof(entry));
+      entry.dwSize = sizeof(entry);
+
+      if (!Process32First(snapshot, &entry))
+        break;
+
+      do
+        {
+          for (i = len - 1; i >= 0; i--)
+            {
+              if (pids[i] == entry.th32ProcessID)
+                break;
+              if (pids[i] == entry.th32ParentProcessID)
+                pids[len++] = entry.th32ProcessID;
+            }
+        }
+      while (len < max_len && Process32Next(snapshot, &entry));
+
+      if (orig_len == len || len >= max_len)
+        break;
+    }
+
+  for (i = len - 1; i >= 0; i--)
+    {
+      HANDLE process = i == 0 ? main_process :
+        OpenProcess(PROCESS_TERMINATE, FALSE, pids[i]);
+
+      if (process)
+        {
+          if (!TerminateProcess(process, exit_code))
+            ret = -1;
+          CloseHandle(process);
+        }
+    }
+
+  return ret;
+}
+
+/**
+ * Inject a thread into the given process that runs ExitProcess().
+ *
+ * Note: as kernel32.dll is loaded before any process, the other process and
+ * this process will have ExitProcess() at the same address.
+ *
+ * This function expects the process handle to have the access rights for
+ * CreateRemoteThread(): PROCESS_CREATE_THREAD, PROCESS_QUERY_INFORMATION,
+ * PROCESS_VM_OPERATION, PROCESS_VM_WRITE, and PROCESS_VM_READ.
+ *
+ * If this method fails, we fall back to running terminate_process_tree().
+ */
+static int
+exit_process(HANDLE process, int exit_code)
+{
+  DWORD code;
+
+  if (GetExitCodeProcess (process, &code) && code == STILL_ACTIVE)
+    {
+      HINSTANCE kernel32 = GetModuleHandle("kernel32");
+      LPTHREAD_START_ROUTINE exit_process_address =
+        (LPTHREAD_START_ROUTINE) GetProcAddress (kernel32, "ExitProcess");
+      DWORD thread_id;
+      HANDLE thread = CreateRemoteThread (process, NULL, 0,
+                                          exit_process_address,
+                                          (PVOID)exit_code, 0, &thread_id);
+
+      if (!thread)
+        return terminate_process_tree(process, exit_code);
```

Comment:
With `terminate_process_tree` the entire process tree is enumerated, and terminated in order.

With `exit_process` we're only signaling the signaling the single process to exit, this could be dangerous.

Per MSDN:

> Exiting a process does not cause child processes to be terminated.